### DISCUSSION TITLE
Job allocator redesign proposal

### DISCUSSION THREAD

**Discussion by jedi5218** [unitystation/unitystation#8423] [Ideas]
### Description
As it is right now, the job allocation algorithm works poorly for very small crews, and in most cases will assign the first job in the list, if they have the same priority. I propose an improved algorithm, that will, hopefully, solve both of those issues.
The algorithm will have a critical job queue, which it will try to assign crew to in order, like:

1. Captain
2. CE
3. CMO
4. QM
5. HoS
6. Chef
...

 Each assignment in this queue must be then assigned in that order, and for that, the algorithm picks crew at random out of these pools:

1. anyone who has this job with high priority, then medium, then low
2. anyone who has a high priority in this department, then medium, then low
3. everyone else

If no player was found in the first pool, the algorithm attempts to pick a player out of the next pool, until at least one qualifying player is found. This way, for example, a single player will always be assigned Captain, and the crew of 5 will have every head of departments, which should mean the crew would never be left with no access to medical, or no engineer to fill the plasma generators.
When all the critical jobs are assigned, the rest of the players can be assigned as follows:

1. Shuffle the remaining player list
2. for each player out of the queue:
    Try to assign him to the job he has set at a high priority. If that position is already taken, try to assign him to a random job out 
    of his medium priority, then, once those run out, to a random job with the low priority, then to Assistant, if all of the low 
    priority jobs are also filled.

**Comment by Fogapod** [comment]
Should allocator have randomness or same people should consistently be picked for same jobs when play together? That's a bit boring

**Comment by jedi5218** [comment]
it should definitely have randomness, and this is exactly the problem with the current algorithm. As it is going through the occupations in the set order, it will most often assign the same players to the same position. For example, if you have every job set to medium, you will almost always be assigned QM, because it's the first head position on the list unless it is contested. This means, that with the current algorithm not only will players be assigned to a one set job in the majority of cases, it will also mean that most shifts with small crew start without any command. This is exactly the reason why I am proposing the redesign, with the random player assignment and guaranteed filling of dept head positions

**Comment by wraith-54321** [comment]
we might also want to add some kind of general job groups, as I have found that in BYOND I will get a head of staff 5 times in a row when I have it turned off, even though there are 2 or 3 other heads of staff, maybe it tries to get X amount of jobs filled in each department before exact jobs(with a few exceptions)

**Comment by AutumnNova** [comment]
this will result in people getting picked for roles that they wont want to play, in which case their is a decent chance they just disconnect, you know that, right?

**Comment by Fogapod** [comment]
Personally, I feel more frustrated when I get picked as CE 10 times in a row, being given random role once in a while isn't that big of a deal? Just make sure you avoid giving people roles randomly if setting is on "low"

**Comment by jedi5218** [comment]
with the current system, I tend to wait for the round to start to then pick the job that's needed because otherwise I will always be selected as a bartender. 

> this will result in people getting picked for roles that they won't want to play, in which case there is a decent chance they just disconnect, you know that, right?

in order to get a job that was not in your job preferences at all, a lot of things need to happen:
1. out of all the crew, no one has set a preference for this critical job
2. no one has any settings in the department
3. you get assigned this job at random because absolutely no one shows any affinity for this job

and I think that having a very small chance of being assigned to the job you don't have a preference for is worth having a functional crew every shift.

**Comment by AutumnNova** [comment]
> with the current system, I tend to wait for the round to start to then pick the job that's needed because otherwise I will always be selected as a bartender.
> 
> > this will result in people getting picked for roles that they won't want to play, in which case there is a decent chance they just disconnect, you know that, right?
> 
> in order to get a job that was not in your job preferences at all, a lot of things need to happen:
> 
>     1. out of all the crew, no one has set a preference for this critical job
> 
>     2. no one has any settings in the department
> 
>     3. you get assigned this job at random because absolutely no one shows any affinity for this job
> 
> 
> and I think that having a very small chance of being assigned to the job you don't have a preference for is worth having a functional crew every shift.

this will clash with an hour limit on jobs
and having someone who doesnt know or doesnt want to play their job isn't much better then noone playing that job at all

**Comment by AutumnNova** [comment]
> Personally, I feel more frustrated when I get picked as CE 10 times in a row, being given random role once in a while isn't that big of a deal? Just make sure you avoid giving people roles randomly if setting is on "low"

if im understanding this correctly, setting preference to "none" gives you a higher chance then low?
seems extremely couner intuitive

**Comment by jedi5218** [comment]
> > with the current system, I tend to wait for the round to start to then pick the job that's needed because otherwise I will always be selected as a bartender.
>  ...

> this will clash with an hour limit on jobs
> and having someone who doesnt know or doesnt want to play their job isn't much better then noone playing that job at all

Right now there is no playtime limit at all, AFAIR, and with the proposed system it should be fairly easy to remove players without necessary playtime from the candidate pools. It is also possible to just remove the last pool (giving the job to anyone if no candidate is found), and the algorithm would still work better than what is implemented now. In my opinion, however, it is better to have someone who is able to at least surrender his duties, rather than having to hack into the captain's office for a spare card.

**Comment by AutumnNova** [comment]
job bans and time requirements would reintroduce the possibility for job slots to go unfilled, besides, the chain of command exists for a reason. If no capt exists a head will be temp promoted to cap.

**Comment by jedi5218** [comment]
> job bans and time requirements would reintroduce the possibility for job slots to go unfilled, besides, the chain of command exists for a reason. If no capt exists a head will be temp promoted to cap.

right now there is no mechanism in place for such a promotion (i.e. without HoP or Captain to make it), which means that to get "promoted", you have to hack your way to the captain's spare id.

**Comment by NoooneyDude** [comment]
Related to issue #5192.
